### PR TITLE
Feat issue 1815

### PR DESCRIPTION
**Pull request by chuntaojun**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

feat issue #1815 #1823 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by nkorange** [comment]
1. 可以参考Dubbo客户端的系统钩子机制
2. 注意start和destroy的幂等性

File: api/src/main/java/com/alibaba/nacos/api/DestructionFactory.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ **/
+public class DestructionFactory {
+
+    public static void registerShutdownHook(final LifeCycle lifeCycle) {
+        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    lifeCycle.destroy();
+                    // just for test
```
Comment (nkorange) [inline api/src/main/java/com/alibaba/nacos/api/DestructionFactory.java]: 这行注释？

File: api/src/main/java/com/alibaba/nacos/api/LifeCycle.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * @date 2019-09-06 15:07
+ **/
+public interface LifeCycle {
+
+    /**
+     * The success of start-up
+     *
+     * @return start label
+     */
+    boolean isStart();
```
Comment (nkorange) [inline api/src/main/java/com/alibaba/nacos/api/LifeCycle.java]: 改成isStarted，下同

File: test/src/test/java/com/alibaba/nacos/test/lifecycle/LifeCycle_ITCase.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    @Test
+    public void destructionService() throws NacosException {
+
+        String serverAddr = "127.0.0.1:" + port;
+
+        namingMaintainService = NacosFactory.createMaintainService(serverAddr);
+        namingService = NamingFactory.createNamingService(serverAddr);
+        configService = NacosFactory.createConfigService(serverAddr);
+
+        Assert.assertTrue(namingMaintainService.isDestroy());
```
Comment (nkorange) [inline test/src/test/java/com/alibaba/nacos/test/lifecycle/LifeCycle_ITCase.java]: return true?

**Comment by nkorange** [comment]
And fix the CI error please.

**Comment by chuntaojun** [comment]
ok, got it

**Pull request closed without merging**